bpo-11063: Add a configure check for uuid_generate_time_safe

berkerpeksag opened the pull request:
<!-- issue-number: bpo-11063 -->
https://bugs.python.org/issue11063
<!-- /issue-number -->

berkerpeksag committed:
bpo-11063: Add a configure check for uuid_generate_time_safe

pitrou commented on pyconfig.h.in:
```diff
@@ -571,6 +571,9 @@
 /* Define to 1 if you have the <libutil.h> header file. */
 #undef HAVE_LIBUTIL_H
 
+/* Define to 1 if you have the `uuid' library (-luuid). */
```
It's not the uuid library, it's specifically  uuid_generate_time_safe() (which exists on Linux, but not on OS X, even though OS X *does* have libuuid).

skrah replied:
That looks autogenerated.  The configure check works here on Linux.

With the patch:

   `checking for uuid_generate_time_safe in -luuid... yes`

Now use a non-existing function name:

  `AC_CHECK_LIB(uuid, uuid_generate_time_unsafe)`

  `checking for uuid_generate_time_unsafe in -luuid... no`


I did not look at the autogenerated `configure`, but it seems to work here.

pitrou commented on configure:
```diff
@@ -9511,6 +9511,51 @@ _ACEOF
+cat confdefs.h - <<_ACEOF >conftest.$ac_ext
+/* end confdefs.h.  */
+
+/* Override any GCC internal prototype to avoid an error.
```
Is this comment copy-pasted from somewhere else? It doesn't seem to apply here...

pitrou commented on configure:
```diff
@@ -9511,6 +9511,51 @@ _ACEOF
+#ifdef __cplusplus
+extern "C"
+#endif
+char uuid_generate_time_safe ();
```
The actual signature here is `int uuid_generate_time_safe(uuid_t out)`.

pitrou commented on configure:
```diff
@@ -9511,6 +9511,51 @@ _ACEOF
+int
+main ()
+{
+return uuid_generate_time_safe ();
```
This call should match the function's signature.

pitrou commented on configure:
```diff
@@ -9511,6 +9511,51 @@ _ACEOF
+  $as_echo_n "(cached) " >&6
+else
+  ac_check_lib_save_LIBS=$LIBS
+LIBS="-luuid  $LIBS"
```
`setup.py` has a bit of code to find `libuuid` in the proper place (`self.compiler.find_library_file(lib_dirs, 'uuid')`). It sounds like you should replicate that in the configure script, but it looks like you don't?

pitrou commented on configure:
```diff
@@ -9511,6 +9511,51 @@ _ACEOF
+#define HAVE_LIBUUID 1
+_ACEOF
+
+  LIBS="-luuid $LIBS"
```
I'm not sure about this. `-luuid` is only needed for the `_uuid` extension module, and it's already added by `setup.py`.

berkerpeksag committed:
another attempt

berkerpeksag committed:
this should work

vstinner commented on Modules/_uuidmodule.c:
```diff
@@ -4,17 +4,10 @@
 py_uuid_generate_time_safe(void)
 {
-#ifdef HAVE_TIME_SAFE
+#if defined(HAVE_UUID_GENERATE_TIME_SAFE) && HAVE_UUID_GENERATE_TIME_SAFE
```
I don't think that it's correct to test the define *value*. Only check if it's defined:

#ifdef HAVE_UUID_GENERATE_TIME_SAFE

berkerpeksag replied:
Yeah, I was about to write about that. I couldn't make autoconf set `#define HAVE_UUID_GENERATE_TIME_SAFE 0` if `uuid_generate_time_safe()` isn't available yet. Working on it :)

skrah replied:
What was broken about the original patch except that some autogenerated comments were suboptimal?

The existing `AC_CHECK_LIB(dld, shl_load)` should have all the same problems.

berkerpeksag committed:
address review comments

berkerpeksag commented:
Ok, I think I've addressed most of Antoine's (and Victor's) comments:

* Renamed `HAVE_LIBUUID` to `HAVE_UUID_GENERATE_TIME_SAFE`
* Don't set `-luuid` in `LIBS`. `setup.py` already covers this.
* Switched back to use `#ifdef HAVE_UUID_GENERATE_TIME_SAFE` in `_uuidmodule.c`
* The `uuid_generate_time_safe` signature in the test program now matches the actual implementation
* Removed the `/* Override any GCC internal prototype to avoid an error. */` comment

vstinner commented on configure.ac:
```diff
@@ -2677,6 +2677,19 @@ AC_CHECK_LIB(sendfile, sendfile)
+        uuid_t out;
+        uuid_generate_time_safe(out);
+        ])],
+    [ax_cv_uuid_generate_time_safe=yes],
```
Is it possible to put directly the AC_DEFINE() here, instead of using a variable?

[ax_cv_uuid_generate_time_safe=no] would be replaced with [] in that case.

berkerpeksag replied:
I tried that but I got the following warning:

```
configure.ac:2688: warning: AC_CACHE_VAL(ax_cv_uuid_generate_time_safe, ...): suspicious presence of an AC_DEFINE in the second argument, where no actions should be taken
```

vstinner replied:
Oh, I hate autotools, mysterious tools... Ok, leave it as it is ;-)

vstinner reviewed (approved):
LGTM.

Thank you Berker, you are more brave than me :-)

berkerpeksag committed:
simplify check a bit
